ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/RecursivePeerRefreshState.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright 2018 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.p2p.discovery.internal;
+
+import static java.util.stream.Collectors.toList;
+import static tech.pegasys.pantheon.ethereum.p2p.discovery.internal.PeerDistanceCalculator.distance;
+
+import tech.pegasys.pantheon.ethereum.p2p.peers.Peer;
+import tech.pegasys.pantheon.ethereum.p2p.peers.PeerBlacklist;
+import tech.pegasys.pantheon.util.bytes.BytesValue;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import io.vertx.core.Vertx;
+
+class RecursivePeerRefreshState {
+  private final int CONCURRENT_REQUEST_LIMIT = 3;
+  private final int TIMEOUT_TASK_DELAY = 30000; // 30 Seconds
+  private final BytesValue target;
+  private final PeerBlacklist peerBlacklist;
+  private final BondingAgent bondingAgent;
+  private final NeighborFinder neighborFinder;
+  private final List<PeerDistance> anteList;
+  private final List<OutstandingRequest> outstandingRequestList;
+  private final List<BytesValue> contactedInCurrentExecution;
+  private final Vertx vertx;
+
+  RecursivePeerRefreshState(
+      final BytesValue target,
+      final PeerBlacklist peerBlacklist,
+      final BondingAgent bondingAgent,
+      final NeighborFinder neighborFinder,
+      final Vertx vertx) {
+    this.target = target;
+    this.peerBlacklist = peerBlacklist;
+    this.bondingAgent = bondingAgent;
+    this.neighborFinder = neighborFinder;
+    this.vertx = vertx;
+    this.anteList = new ArrayList<>();
+    this.outstandingRequestList = new ArrayList<>();
+    this.contactedInCurrentExecution = new ArrayList<>();
+
+    commenceTimeoutTask();
+  }
+
+  private void commenceTimeoutTask() {
+    vertx.setPeriodic(
+        TIMEOUT_TASK_DELAY,
+        v -> {
+          List<OutstandingRequest> outstandingRequestListCopy =
+              new ArrayList<>(outstandingRequestList);
```

Comment:
Why are we making a copy here? It doesn't provide any thread safety because making the copy requires iterating the list.

For thread safety through this class, I suspect we'll need to use `synchronized` blocks so that only one thread can be anywhere in this class at a time.